lib/src/equality.dart

```diff
@@ -32,34 +32,44 @@ abstract class Equality<E> {
 
 typedef F _GetKey<E, F>(E object);
 
-/// Equality of objects using a canonical value of the object itself.
+/// Equality of objects based on derived values.
 ///
 /// For example, given the class:
+///     ```
 ///     abstract class Employee {
 ///       int get employmentId;
 ///     }
+///     ```
 ///
-/// Creating an [Equality] object that uses `employmentId`:
-///     new EqualityBy<Employee, int>((e) => e.employmentId);
+/// The following [Equality] considers employees with the same IDs to be equal:
+///     ```
+///     new EqualityBy((Employee e) => e.employmentId);
+///     ```
 ///
 /// It's also possible to pass an additional equality instance that should be
 /// used to compare the value itself.
 class EqualityBy<E, F> implements Equality<E> {
+  // Returns a derived value F from an object E.
   final _GetKey<E, F> _getKey;
+
+  // Determines equality between two values of F.
   final Equality<F> _inner;
 
   EqualityBy(F getKey(E object), [Equality<F> inner = const DefaultEquality()])
       : _getKey = getKey,
         _inner = inner;
 
-  @override
   bool equals(E e1, E e2) => _inner.equals(_getKey(e1), _getKey(e2));
 
-  @override
   int hash(E e) => _inner.hash(_getKey(e));
 
-  @override
-  bool isValidKey(Object o) => o is E;
+  bool isValidKey(Object o) {
+    if (o is E) {
+      final value = _getKey(o);
+      return value is F && _inner.isValidKey(value);
```

Comment:
I don't think this `is F` check is necessary, since `_getKey()` is typed to return an `F`. As a general rule, I don't write runtime checks that are unnecessary if user code is strong-mode clean.